### ISSUE TITLE
Avoid or More graceful handling of frequent websocket disconnects.

### ISSUE THREAD

**Issue by mdeweerd** [home-assistant/frontend#17454]
![image](https://github.com/home-assistant/home-assistant-js-websocket/assets/1504752/5b203ed4-9539-4fdf-87bd-2f7b6dd72071)

The above image is a typical output of an event listener tracking the delay between two successive (web-socket) connection state changes (shared below).

For an unknown reason the websocket is interrupted every 30+ seconds and reconnects in less than 70ms.

That causes a reload of page items, which include the floorplan which is then initialised to the initial state (view).

I tracked back this reload to the "connection-status" event listener that is added in the code below in ha-panel-lovelace.js:

```javascript
  protected firstUpdated(changedProps) {
    super.firstUpdated(changedProps);

    this._fetchConfig(false);
    if (!this._unsubUpdates) {
      this._subscribeUpdates();
    }
    // reload lovelace on reconnect so we are sure we have the latest config
    window.addEventListener("connection-status", (ev) => {
      if (ev.detail === "connected") {
        this._fetchConfig(false);
      }
    });
  }
```

To track the changes in the connection state and the delay between the changes, I used the next code:
```javascript
window.addEventListener("connection-status", (ev) => {
    var n=Date.now();var d='?';
    if(typeof(window.pConStateTime)!=='undefined') {
        d=n-window.pConStateTime;
    }
    console && console.log({"elapsed":d,"detail":ev.detail+""});
    window.pConStateTime=n;
    });
```

It occurs with Chrome and Firefox.  It occurs on different computers.
When I reload the pages on two different browsers at about the same time, the pages are refreshed at about the same time.  Otherwise they are refreshed at different times.
It does not seem to happen on a "busy" page.

In my search for a solution, I found some similar issues in other projects:
- [Disconnects and reconnects every 30 seconds](https://github.com/yjs/y-websocket/issues/24) - but in my case it does not matter whether I am debugging or not.
- [How to fix WebSockets/HAProxy timeout](https://lucjan.medium.com/investigating-websocket-haproxy-reconnecting-and-timeouts-6d19cc0002a1) - I am using ha-proxy, so that may cause the disconnects.  This article also seems to suggest that reducing the delays between heartbeats helps.  I modified two delays in the HAProxy settings (client timeout and keepalive timeout) but the disconnects still happened after 30s.
![image](https://github.com/home-assistant/home-assistant-js-websocket/assets/1504752/9dc280a4-c80e-42e4-b57f-bc36ee4c9eb9)

EDIT: By adding an automation ensuring a state update every 25s impacting the floorplan, the websocket was not lost.

**Comment by balloob** [comment]
Do you use something like nginx inbetween that closes connections? Neither browsers or Home Assistant automatically close websocket connections.

**Comment by mdeweerd** [comment]
I use haproxy (in OPNSense, my firewall) - I tweaked some settings there, but that did not help.
I do not know if there currently is a "keep-alive" mechanism, but that would surely fix the issue.
What helps is having a virtual light change on/off state every 25 seconds.

**Comment by balloob** [comment]
This sounds like an issue with your haproxy configuration. Direct Home Assistant operation doesn't suffer from this.

**Comment by mdeweerd** [comment]
The nginx add on also suffers from this.

It is clear that the proxy interrupts the connection, and IMHO the solution is to send keep-alive packets as suggested by some of the resources I found.

To confirm the issue with nginx, I tested this with another setup where HAOS is running on a raspberry pi with the nginx add-on serving as the https proxy sitting directly behind an internet box with public ports 80 and 443 natted to the raspberry pi.  The same happens there but one has to wait for about 15 minutes - I guess that the timeout value is 900s in nginx.

This is the screenshot of the websocket timing in the Developer Tools view of Chrome:
![image](https://github.com/home-assistant/frontend/assets/1504752/bbf5c26c-dea3-4bd2-9038-3f9e8e82c95d).

Increased timeouts in the proxies would defeat IMHO their main purpose (limit the number of abandonned but apparently open connections).
